File: Algorithms/Python/mergesort.py
def merge(left, right):
    li, ri = 0, 0
    retval = []
    while li < len(left) and ri < len(right):
        if left[li] < right[ri]:
            retval.append(left[li])
            li += 1
        else:
            retval.append(right[ri])
            ri += 1

    retval += left[li:]
    retval += right[ri:]
    return retval


def merge_sort(array):
    # base case
    if len(array) <= 1:  
        return array

    # divide array in half and do recursion
    half = len(array) // 2
    left = merge_sort(array[:half])
    right = merge_sort(array[half:])

    return merge(left, right)


def sort():
    list = [5, 7, 6, 2, 1, 7, 3]
    sorted_list = merge_sort(list)
    print 'Array after sorting: ', sorted_list

if __name__ == '__main__':
    sort()

File: Algorithms/Java/MergeSort.java
void merge(int array_[], int left, int middle, int right) 
    { 
        int n1 = middle - left + 1; 
        int n2 = right - middle; 
        int lhs[] = new int [n1]; 
        int rhs[] = new int [n2]; 

        for (int i=0; i<n1; ++i) 
            lhs[i] = array_[left + i]; 
        for (int j=0; j<n2; ++j) 
            rhs[j] = array_[middle + 1+ j]; 

        int i = 0, j = 0; 
        int k = left; 
        while (i < n1 && j < n2) 
        { 
            if (lhs[i] <= rhs[j]) 
            { 
                array_[k] = lhs[i]; 
                i++; 
            } 
            else
            { 
                array_[k] = rhs[j]; 
                j++; 
            } 
            k++; 
        } 
        while (i < n1) 
        { 
            array_[k] = lhs[i]; 
            i++; 
            k++; 
        } 
        while (j < n2) 
        { 
            array_[k] = rhs[j]; 
            j++; 
            k++; 
        } 
    } 
    void sort(int array_[], int left, int right) 
    { 
        if (left < right) 
        { 
            int middle = (left+right)/2; 
            sort(array_, left, middle); 
            sort(array_ , middle+1, right);
            merge(array_, left, middle, right); 
        } 
    }